### PR TITLE
Fixed dependency issue when SecurityBundle is not installed

### PR DESCRIPTION
**Pull request by reypm**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | https://github.com/api-platform/api-platform/issues/1265
| License       | MIT

### CODE REVIEW

**Comment by alanpoulain** [comment]
You need to modify the code inside the stages too to handle the case when `$resourceAccessChecker` is null.

**Comment by reypm** [comment]
> You need to modify the code inside the stages too to handle the case when `$resourceAccessChecker` is null.

What do you mean? Not following you on the word `stages` 🤔

**Comment by alanpoulain** [comment]
The services you are modifying are "stages" (`SecurityStage`, `SecurityPostDenormalizeStage`).

**Comment by reypm** [comment]
> The services you are modifying are "stages" (`SecurityStage`, `SecurityPostDenormalizeStage`).

You mean add an extra check at for `null` check at `src/GraphQl/Resolver/Stage/SecurityStage.php`??

```
    if (null === $isGranted || null === $this->resourceAccessChecker || $this->resourceAccessChecker->isGranted($resourceClass, (string) $isGranted, $context['extra_variables'])) {
        return;
    }
```

**Comment by alanpoulain** [comment]
Even before this condition but that's the idea. And make the variable in the constructor nullable.
Don't forget to add unit tests too!

**Comment by teohhanhui** [comment]
@reypm

```php
    if (null === $isGranted || null === $this->resourceAccessChecker || $this->resourceAccessChecker->isGranted($resourceClass, (string) $isGranted, $context['extra_variables'])) {
        return;
    }
```

This is wrong. We must not silently allow access if the user has specified some condition (i.e. `$isGranted` is not `null`) but `$this->resourceAccessChecker` is `null`. We should fail in this case.

**Comment by alanpoulain** [comment]
Oh yes you're right @teohhanhui

**Comment by reypm** [comment]
> We should fail in this case.

@teohhanhui and by fail you mean throw the Error as in last line on that `__invoke` function? I am following you right you want this? 

```
/**  @var ResolveInfo $info */
$info = $context['info'];

$isGranted = $resourceMetadata->getGraphqlAttribute($operationName, 'security', null, true);
if (null === $this->resourceAccessChecker && $isGranted) {
    throw Error::createLocatedError($resourceMetadata->getGraphqlAttribute($operationName, 'security_message', 'Access Denied.'), $info->fieldNodes, $info->path);
}

if (null === $isGranted || $this->resourceAccessChecker->isGranted($resourceClass, (string) $isGranted, $context['extra_variables'])) {
    return;
}

throw Error::createLocatedError($resourceMetadata->getGraphqlAttribute($operationName, 'security_message', 'Access Denied.'), $info->fieldNodes, $info->path);
```

@alanpoulain yes, I will add the test once the desired solution is approved by you both :) 

Sorry for not being following you guys, first time "fixing" an api-platform issue and understanding it's internal.

**Comment by teohhanhui** [comment]
> @teohhanhui and by fail you mean throw the Error as in last line on that __invoke function? I am following you right you want this?

I think we should do something like:

```php
throw new \LogicException('Cannot check security expression when SecurityBundle is not installed. Try running "composer require symfony/security-bundle".');
```

as this is a configuration issue.

Or do you always want to avoid throwing exceptions at all costs, @alanpoulain? (Why?)

**Comment by alanpoulain** [comment]
We can throw a `LogicException`, it will be converted in an `Error` when https://github.com/api-platform/core/pull/3063 will be merged (and it's not an issue in the meantime if it's not the case).

**Comment by reypm** [comment]
@teohhanhui / @alanpoulain better now? Did I miss something else?

File: src/GraphQl/Resolver/Stage/SecurityPostDenormalizeStage.php [hunk 1]
```diff
@@ -52,7 +57,7 @@ public function __invoke(string $resourceClass, string $operationName, array $co
             }
         }
 
-        if (null === $isGranted || $this->resourceAccessChecker->isGranted($resourceClass, (string) $isGranted, $context['extra_variables'])) {
+        if (null === $isGranted || null === $this->resourceAccessChecker || $this->resourceAccessChecker->isGranted($resourceClass, (string) $isGranted, $context['extra_variables'])) {
```
Comment (teohhanhui) [inline src/GraphQl/Resolver/Stage/SecurityPostDenormalizeStage.php]: This change is unnecessary. It's not possible at this point that `null !== $isGranted && null === $this->resourceAccessChecker` would evaluate to `true` (exactly what we've already checked above). (I'd also suggest changing the above condition to this form, as it reads better.)

Comment (reypm) [reply src/GraphQl/Resolver/Stage/SecurityPostDenormalizeStage.php]: Ups! That was my bad I miss that check in there, fixing it now.

File: src/Bridge/Symfony/Bundle/Resources/config/graphql.xml [hunk 2]
```diff
@@ -53,12 +53,12 @@
 
         <service id="api_platform.graphql.resolver.stage.security" class="ApiPlatform\Core\GraphQl\Resolver\Stage\SecurityStage" public="false">
             <argument type="service" id="api_platform.metadata.resource.metadata_factory" />
-            <argument type="service" id="api_platform.security.resource_access_checker" />
+            <argument type="service" id="api_platform.security.resource_access_checker" on-invalid="ignore" />
```
Comment (dunglas) [inline src/Bridge/Symfony/Bundle/Resources/config/graphql.xml]: Should be `null`, not `ignore` isn't it? Or calling the constructor will still fail.

Comment (reypm) [reply src/Bridge/Symfony/Bundle/Resources/config/graphql.xml]: Why test are passing with this configuration? 🤔

Comment (dunglas) [reply src/Bridge/Symfony/Bundle/Resources/config/graphql.xml]: Because the security component is installed on Travis.

Comment (teohhanhui) [reply src/Bridge/Symfony/Bundle/Resources/config/graphql.xml]: No, they behave the same way. The only difference is when inside a method call.

https://symfony.com/doc/current/service_container/optional_dependencies.html#ignoring-missing-dependencies

Comment (teohhanhui) [reply src/Bridge/Symfony/Bundle/Resources/config/graphql.xml]: I suggest for consistency's sake, we should not use `on-invalid="null"` except in the specific cases where we need that behaviour (i.e. a method call with `null` value).

File: src/GraphQl/Resolver/Stage/SecurityStage.php [hunk 3]
```diff
@@ -44,6 +44,10 @@ public function __invoke(string $resourceClass, string $operationName, array $co
         $resourceMetadata = $this->resourceMetadataFactory->create($resourceClass);
 
         $isGranted = $resourceMetadata->getGraphqlAttribute($operationName, 'security', null, true);
+        if (null === $this->resourceAccessChecker && $isGranted) {
```
Comment (alanpoulain) [inline src/GraphQl/Resolver/Stage/SecurityStage.php]: Please add a new line before?

File: src/GraphQl/Resolver/Stage/SecurityPostDenormalizeStage.php [hunk 4]
```diff
@@ -44,6 +44,11 @@ public function __invoke(string $resourceClass, string $operationName, array $co
         $resourceMetadata = $this->resourceMetadataFactory->create($resourceClass);
 
         $isGranted = $resourceMetadata->getGraphqlAttribute($operationName, 'security_post_denormalize', null, true);
+
+        if (null === $this->resourceAccessChecker && $isGranted) {
+            throw new \LogicException('Cannot check security expression when SecurityBundle is not installed. Try running "composer require symfony/security-bundle".');
+        }
+
```
Comment (alanpoulain) [inline src/GraphQl/Resolver/Stage/SecurityPostDenormalizeStage.php]: I think we have an issue if `null === $this->resourceAccessChecker` and `$isGranted` is false (see the code below).

File: tests/GraphQl/Resolver/Stage/SecurityPostDenormalizeStageTest.php [hunk 5]
```diff
@@ -115,4 +115,34 @@ public function testNotGranted(): void
...
     }
+
+    public function testNoSecurityBundleInstalled(): void
+    {
+        $operationName = 'item_query';
+        $resourceClass = 'myResource';
+        $isGranted = 'not_granted';
+        $extraVariables = ['extra' => false];
+        $resourceMetadata = (new ResourceMetadata())->withGraphql(
+            [
```
Comment (alanpoulain) [inline tests/GraphQl/Resolver/Stage/SecurityPostDenormalizeStageTest.php]: You can follow the other CS.

File: tests/GraphQl/Resolver/Stage/SecurityStageTest.php [hunk 6]
```diff
@@ -96,4 +96,18 @@ public function testNotGranted(): void
...
+        $this->securityStage = new SecurityStage($this->resourceMetadataFactoryProphecy->reveal(), null);
+
+        $operationName = 'item_query';
+        $resourceClass = 'myResource';
+        $isGranted = 'not_granted';
+        $extraVariables = ['extra' => false];
+        $resourceMetadata = (new ResourceMetadata())->withGraphql([$operationName => ['security' => $isGranted]]);
+        $this->resourceMetadataFactoryProphecy->create($resourceClass)->willReturn($resourceMetadata);
+        $this->expectException(\LogicException::class);
+        ($this->securityStage)($resourceClass, 'item_query', ['extra_variables' => $extraVariables]);
```
Comment (alanpoulain) [inline tests/GraphQl/Resolver/Stage/SecurityStageTest.php]: Why the CS is so different (new lines, ...)?

File: src/GraphQl/Resolver/Stage/SecurityPostDenormalizeStage.php [hunk 7]
```diff
@@ -44,6 +44,11 @@ public function __invoke(string $resourceClass, string $operationName, array $co
         $resourceMetadata = $this->resourceMetadataFactory->create($resourceClass);
 
         $isGranted = $resourceMetadata->getGraphqlAttribute($operationName, 'security_post_denormalize', null, true);
+
+        if (null === $this->resourceAccessChecker && $isGranted) {
```
Comment (teohhanhui) [inline src/GraphQl/Resolver/Stage/SecurityPostDenormalizeStage.php]:
```suggestion
        if (null !== $isGranted && null === $this->resourceAccessChecker) {
```

File: src/GraphQl/Resolver/Stage/SecurityStage.php (hunk 3 above)
Comment (teohhanhui) [inline src/GraphQl/Resolver/Stage/SecurityStage.php]:
```suggestion
        if (null !== $isGranted && null === $this->resourceAccessChecker) {
```

File: tests/GraphQl/Resolver/Stage/SecurityStageTest.php (hunk 6 above)
Comment (reypm) [reply tests/GraphQl/Resolver/Stage/SecurityStageTest.php]: That was maybe my IDE setup for CS, I will fix it in the next commit :)

File: src/GraphQl/Resolver/Stage/SecurityStage.php (hunk 3 above)
Comment (reypm) [reply src/GraphQl/Resolver/Stage/SecurityStage.php]: This change has been applied, thanks

File: tests/GraphQl/Resolver/Stage/SecurityPostDenormalizeStageTest.php (hunk 5 above)
Comment (reypm) [reply tests/GraphQl/Resolver/Stage/SecurityPostDenormalizeStageTest.php]: Done.

File: src/GraphQl/Resolver/Stage/SecurityPostDenormalizeStage.php (hunk 4 above)
Comment (reypm) [reply src/GraphQl/Resolver/Stage/SecurityPostDenormalizeStage.php]: Fixed with @teohhanhui suggestions.

**Comment by alanpoulain** [comment]
I've rebased your PR for 2.5 and made some little changes. Thank you @reypm!

**Pull request merged**